security/pkg/server/ca/node_auth.go

```diff
@@ -16,27 +16,157 @@ package ca
 // valid identities which run on that node (rather than arbitrary ones).
-type NodeAuthorizer struct {
+type NodeAuthorizer interface {
+	authenticateImpersonation(caller security.KubernetesInfo, requestedIdentityString string) error
+}
+
+// MulticlusterNodeAuthorizor is an implementation of NodeAuthorizer for multi-cluster environmnets.
+// It should be used when ENABLE_MULTICLUSTER_NODE_AUTHORIZER is set, which means node authorizations
+// from remote clusters could also be processed (https://docs.google.com/document/d/10uf4EvUVif4xGeCYQydaKh9Yaz9wpysao7gyLewJY2Q)
+// It is responsible for maintaining an index of node authenticators, one per cluster.
+// node authorizations from one cluster will be forwarded to the node authorizer for the same cluster.
+type MulticlusterNodeAuthorizor struct {
+	remoteNodeAuthenticators map[cluster.ID]*ClusterNodeAuthorizer
+	ztunnelPodsClient        map[cluster.ID]kclient.Client[*v1.Pod]
+	ztunnelIndex             sync.Map
+	m                        sync.Mutex // protects remoteKubeControllers
+	filter                   func(t any) bool
+	trustedNodeAccounts      map[types.NamespacedName]struct{}
+}
+
+func NewMulticlusterNodeAuthenticator(filter func(t any) bool, trustedNodeAccounts map[types.NamespacedName]struct{},
+	addClusterHandler func(multicluster.ClusterHandler),
+) *MulticlusterNodeAuthorizor {
+	m := &MulticlusterNodeAuthorizor{
+		remoteNodeAuthenticators: map[cluster.ID]*ClusterNodeAuthorizer{},
+		ztunnelPodsClient:        map[cluster.ID]kclient.Client[*v1.Pod]{},
+		filter:                   filter,
+		trustedNodeAccounts:      trustedNodeAccounts,
+	}
+	if addClusterHandler != nil {
+		addClusterHandler(m)
+	}
+	return m
+}
+
+func (mNa *MulticlusterNodeAuthorizor) ClusterAdded(cluster *multicluster.Cluster, _ <-chan struct{}) {
+	na, err := NewClusterNodeAuthorizer(cluster.Client, mNa.filter, mNa.trustedNodeAccounts)
+	if err != nil {
+		serverCaLog.Errorf("failed to initialize node authorizer for cluster %v: %v", cluster.ID, err)
+		return
+	}
+	mNa.m.Lock()
+	defer mNa.m.Unlock()
+	mNa.addCluster(cluster.ID, cluster.Client, na)
+}
+
+func (mNa *MulticlusterNodeAuthorizor) ClusterUpdated(cluster *multicluster.Cluster, stop <-chan struct{}) {
+	na, err := NewClusterNodeAuthorizer(cluster.Client, mNa.filter, mNa.trustedNodeAccounts)
+	if err != nil {
+		serverCaLog.Errorf("failed to initialize node authorizer for cluster %v: %v", cluster.ID, err)
+	}
+	mNa.m.Lock()
+	defer mNa.m.Unlock()
+	mNa.deleteCluster(cluster.ID)
+	mNa.addCluster(cluster.ID, cluster.Client, na)
+}
+
+func (mNa *MulticlusterNodeAuthorizor) ClusterDeleted(key cluster.ID) {
+	mNa.m.Lock()
+	defer mNa.m.Unlock()
+	mNa.deleteCluster(key)
+}
+
+func (mNa *MulticlusterNodeAuthorizor) addCluster(clusterID cluster.ID, client kube.Client, na *ClusterNodeAuthorizer) {
+	ztunnelPodClient := kclient.NewFiltered[*v1.Pod](client, kclient.Filter{
+		ObjectFilter:    isZtunnelPod,
```

Comment:
Can `LabelSelector` be used here?

Reply:
Thanks for your advice! I followed @hzxuzhonghu 's suggestion and use a cluster ID from context to fetch the cluster's node authorizer. Therefore, we don't need to use this "ztunnelPodClient" anymore.